### PR TITLE
Add Provider Functionality

### PR DESCRIPTION
**Pull request by Anaphase**
This PR add a new `Provider` model and replaces the hardcoded `'1'` that was used to denote codex-hosted metadata.

See also:
[web.codex-viewer#53](https://github.com/codex-protocol/web.codex-viewer/pull/53)
[contract.codex-registry#36](https://github.com/codex-protocol/contract.codex-registry/pull/36)

### CODE REVIEW

**Comment by Anaphase** [comment]
NOTE: this PR also moves the "create giveaway" route into a new `/admin` directory, since it was created for the "create provider" route. True admin authentication to come later.

File: src/services/codex-record.js [hunk 1]
```diff
@@ -15,41 +16,68 @@ export default {
...
+
+            // if no provider was found, we can stop here
+            if (!provider) {
+              logger.warn(`Found unknown providerId "${providerId}" for CodexRecord with tokenId ${tokenId}`)
+              return codexRecord
+            }
 
-        return models.CodexRecordMetadata.findById(providerMetadataId)
-          .then((codexRecordMetadata) => {
+            // assocaite the Record with this provider
```
Comment (johnhforrest) [inline src/services/codex-record.js]: typo

File: src/services/codex-record.js [hunk 2]
```diff
@@ -15,41 +16,68 @@ export default {
...
+                if (codexRecordMetadata.codexRecordTokenId !== null) {
+                  throw new Error(`Could not confirm CodexRecord with tokenId ${codexRecord.tokenId} because metadata with id ${providerMetadataId} is already associated with tokenId ${codexRecordMetadata.codexRecordTokenId}.`)
+                }
+
+                codexRecordMetadata.codexRecordTokenId = codexRecord.tokenId
 
-            // @TODO: maybe verify hases here? e.g.:
-            // codexRecordMetadata.nameHash === codexRecord.nameHash
-            // codexRecordMetadata.descriptionHash === codexRecord.descriptionHash
+                // @TODO: maybe verify hases here? e.g.:
```
Comment (johnhforrest) [inline src/services/codex-record.js]: hashes

**Review by johnhforrest — APPROVED** [review]
2 typos

**Review by johnhforrest — APPROVED** [review]

**Pull request merged**